File: web_app/followers/insta.py
from pyppeteer import launch
class Insta():
    def __init__(self, username="", password=""):
        self.__username = username
        self.__password = password
        self.__base_url = "https://instagram.com"
        self.__login_url = self.__base_url
        self.__api_url = self.__login_url+"/{}/?__a=1"

    #Getters
    def getUsername(self):
        return self.__username

    #Setters
    def getPassword(self):
        return self.__password


    async def startup(self):
       self.__browser = await launch({'headless': False,
           'autoClose':False,
           'handleSIGINT':False,
           'handleSIGTERM':False,
           'handleSIGHUP':False})

    async def close(self):
        await self.__browser.close()

    async def login(self):
        username = self.__username
        password = self.__password
        try:
            self.__page = await self.__browser.newPage()
            await self.__page.goto('http://instagram.com')
            await self.__page.waitForSelector('input[name="username"]');
            await self.__page.type('input[name="username"]', username);
            await self.__page.type('input[name="password"]', password);
            await self.__page.click('button[type="submit"]');
            await self.__page.waitForSelector("._6q-tv")
            return True
        except:
            return False

    async def getData(self, username):
        await self.__page.goto(self.__api_url.format(username))
        data = await self.__page.evaluate(('document.body.innerText'))

        followers_start_index = data.find('followed_by\":{\"count\":')
        count = followers_start_index+22
        num_followers=""

        while not data[count]=="}":
            num_followers+=str(data[count])
            count+=1



        following_start_index = data.find('\"edge_follow\"')
        count = following_start_index+23
        num_following = ""

        while not data[count] == "}":
            num_following+=str(data[count])
            count+=1



        profile_pic_start_index = data.find("\"profile_pic_url_hd\"")
        count = profile_pic_start_index+22
        url = ""

        while not data[count] == "\"":
            url += data[count]
            count+=1

        return (num_followers, num_following, url)

File: Insta.js
const puppeteer = require('puppeteer');
const axios = require('axios');

class Insta{
  constructor(username, password){
    this.username = username;
    this.password = password;
    this.baseUrl="https://instagram.com";
  }

  async init(){
    this.browser = await puppeteer.launch({ headless: false }); // default is true
    //this.browser = await puppeteer.launch();
  }

  async login(){
    this.page = await this.browser.newPage();
    await this.page.goto(`${this.baseUrl}/accounts/login/`);
    await this.page.waitForSelector('input[name="username"]');
    await this.page.type('input[name="username"]', this.username);
    await this.page.type('input[name="password"]', this.password);
    await this.page.click('button[type="submit"]');
    await this.page.waitForSelector('._6q-tv');
  }

  async getData(username){
    await this.page.goto(`${this.baseUrl}/${username}/?__a=1`);
    let data = JSON.parse(await this.page.evaluate('document.body.innerText'));



    const numFollowers = data.graphql.user.edge_followed_by.count;
    const numFollowing = data.graphql.user.edge_follow.count;
    const uname = data.graphql.user.username;
    const profilePicURL = data.graphql.user.profile_pic_url;

    const userData = {
      username: uname,
      followers: numFollowers,
      following: numFollowing,
      profilePicURL: profilePicURL
    };


    return userData;
  }

  async close(){
    await this.browser.close();
  }

  async postData(data){
    try{
      const res = await axios.post('http://127.0.0.1:5000/getData',data);
    } catch(err){
      console.log(err);
    }
  }


}


module.exports = Insta;


//Just testing
//ig = new Insta('username', 'password');
//
//(async () => {
//  await ig.init();
//  await ig.login();
//
//
//  await ig.postData(await ig.getData('erickmeyer'));
//  await ig.postData(await ig.getData('amandaaloura'));
//
//
//  await ig.close();
//})(); 
